### PR TITLE
Feature/JC19-370-filter-by-type

### PR DESCRIPTION
**Pull request by khoanguyen-saritasa**

### CODE REVIEW

File: apps/vanilla/src/scripts/renderAnimeList.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          <th class="table__row_item">${element.titleEnglish}</th>
+          <th class="table__row_item">${element.titleJapan}</th>
+          <th class="table__row_item">${element.aired.start ? formatDate(element.aired.start) : ''}</th>
+          <th class="table__row_item">${element.type}</th>
+          <th class="table__row_item">${element.status}</th>
+        </tr>
+      `
+    )).join('');
+
+    if (container) {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderAnimeList.ts:50]: https://wiki.saritasa.rocks/frontend/languages/ts-js/best-practices#dont-use-implicit-type-casting

File: apps/vanilla/src/scripts/renderAnimeList.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    `;
+    }
+    return animeList;
+  } catch (error: unknown) {
+    if (error instanceof Error) {
+      throw new Error(`Unable to render Anime list ${error.message}`);
+    } else {
+      throw new Error('Unexpected error!');
+    }
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderAnimeList.ts:73]: I seen this construction several times. To reduce code duplication, please extract it into a function.

File: apps/vanilla/src/scripts/renderFilterByType.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import { KEY_TYPE } from '../constants/key';
+import { getValueFromLocalStorage, setValueToLocalStorage } from '../service/localStorage';
+
+import { renderListAndPaginationToUI } from './renderPagination';
+
+const selectType = document.querySelector<HTMLSelectElement>('.filter__item_select-type');
+
+/** Render filter by type.
+ * @param options Pagination options.
+ */
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderFilterByType.ts:15]: Please format multiline jsdoc properly, first and last lines shouldn't have text.
```suggestion
/**
 * Render filter by type.
 * @param options Pagination options.
 */
```

File: apps/vanilla/src/scripts/renderFilterByType.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+export function renderFilterByType(options: PaginationOptions): void {
+  const typeOptionHTML = FILTER_TYPE_OPTIONS.map(item => `<option value="${item.title}">${item.title}</option>`).join('');
+  if (selectType === null || selectType === undefined) {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderFilterByType.ts:18]: In our company we prefer to check for undefined and null together, like so
```suggestion
  if (selectType == null) {
```

File: apps/vanilla/src/scripts/renderFilterByType.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    return;
+  }
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderFilterByType.ts:20]: I think you should place this check in the beginning of the function.

File: apps/vanilla/src/scripts/renderPagination.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  const PAGE_RANGE = 5;
+  const PAGE_STEP = 2;
+
+  if (activePage <= PAGE_STEP) {
+    return renderRangeOfPagination(activePage, FIRST_PAGE, PAGE_RANGE);
+  } else if (activePage > PAGE_STEP && activePage < totalPages - PAGE_STEP) {
+    return renderRangeOfPagination(activePage, activePage - PAGE_STEP, activePage + PAGE_STEP);
+  } else if (activePage >= totalPages - PAGE_STEP) {
+    return renderRangeOfPagination(activePage, totalPages - PAGE_STEP * 2, totalPages);
+  }
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderPagination.ts:38]: You don't really need to use `else if` when returning.

File: apps/vanilla/src/scripts/renderPagination.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    });
+
+  buttonLastPage?.addEventListener('click', () => {
+      const optionUpdated = new PaginationOptions({
+        ...options,
+        activePage: options.totalPages,
+        offset: DEFAULT_LIMIT * options.activePage,
+      });
+      renderListAndPaginationToUI(optionUpdated);
+    });
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderPagination.ts:74]: Please fix indentation

File: apps/vanilla/src/scripts/renderPagination.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  } else {
+    buttonFirstPage?.classList.remove('disabled');
+  }
+}
+
+/**
+ * Render anime list and pagination to UI.
+ * @param options Options of pagination.
+ */
+export async function renderListAndPaginationToUI(options: PaginationOptions): Promise<void> {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderPagination.ts:93]: Usually function should do one thing at a time. So having "and" in your function name is a bad practice.
I suggest divide it into two functions.

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    `<option value="${item.title}">${item.title}</option>`
+  )).join('');
+
+  const orderOptionHTML = Object.keys(OrderOption).map(item => (
+    `<option value ="${item}">${item}</option>`
+  ))
+    .join('');
+  if (selectSort === null || selectSort === undefined) {
+    return;
+  }
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:29]: Early return should be placed as early as possible.

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      offset: DEFAULT_LIMIT,
+      activePage: 1,
+      sorting: new Sorting({
+        ...options.sorting,
+        ...selectSortingValue,
+      }),
+    });
+      renderListAndPaginationToUI(optionsUpdated);
+
+    });
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:48]: Please fix indentation

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  }
+  selectOrdering.innerHTML = orderOptionHTML;
+  setDefaultSelected(selectOrdering, getValueFromLocalStorage<OrderOption>(KEY_ORDER) ?? OrderOption.Ascending);
+  selectOrdering.addEventListener('change', () => {
+
+    setValueToLocalStorage(KEY_ORDER, selectOrdering.value);
+    const selectOrderingValue = getValueFromLocalStorage<OrderOption>(KEY_ORDER) ?? OrderOption.Ascending;
+
+    /** Get type of ordering option.*/
+    function getSelectOptions(): boolean {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:61]: You can use just a variable with a ternary operator here.

File: libs/core/dtos/anime.dto.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+export interface AnimeDto {
+
+  /** Anime ID. */
+  readonly id: number;
+
+  /** Date created. */
+  readonly created: number;
+
+  /** Date modified. */
+  readonly modified: string;
```
Comment (artem-sotnichenko-saritasa) [inline libs/core/dtos/anime.dto.ts:15]: It is strange that dates have different types.
If it not your typo, but api bug, consider reporting it in our slack camp chat.

**Review by artem-sotnichenko-saritasa — CHANGES_REQUESTED** [review]

File: libs/core/dtos/anime.dto.ts (hunk 12 above)
Comment (khoanguyen-saritasa) [reply libs/core/dtos/anime.dto.ts:15]: It is type string for created date. This is my mistake.

File: apps/vanilla/src/scripts/renderAnimeList.ts (hunk 2 above)
Comment (khoanguyen-saritasa) [reply apps/vanilla/src/scripts/renderAnimeList.ts:72]: I extracted it into function in feature show anime detail so can I skip the extraction in this feature?

Comment (artem-sotnichenko-saritasa) [reply apps/vanilla/src/scripts/renderAnimeList.ts:72]: @khoanguyen-saritasa I think you can skip it in this PR.

File: apps/vanilla/src/constants/filterTypeOptions.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+
+/** Options of type interface. */
+interface FilterType {
+
+  /** Title of options. */
+  title: string;
+
+  /** Value of options. */
+  value: TypeModel;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/constants/filterTypeOptions.ts:10]: Don't forget to add readonly.

File: apps/vanilla/src/constants/filterTypeOptions.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+import { TypeModel } from '@js-camp/core/enum';
+
+export const FILTER_TYPE_OPTIONS: FilterType[] = [
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/constants/filterTypeOptions.ts:15]: Don't forget to add readonly for arrays too

File: apps/vanilla/src/scripts/initAnimeTable.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import { renderListAnimeWithActivePage } from './renderPagination';
+import { renderSortingAndOrdering } from './renderSortingAndOrdering';
+
+/** Init anime table view. */
+export async function initAnimeTable(): Promise<void> {
+  const INITIAL_PAGINATION: PaginationOptions = new PaginationOptions({
+    limit: DEFAULT_LIMIT,
+    offset: DEFAULT_OFFSET,
+    activePage: DEFAULT_ACTIVE_PAGE,
+    totalPages: 0,
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/initAnimeTable.ts:20]: Be consistent and use constant as you do above.

File: apps/vanilla/src/scripts/renderFilterByType.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+ */
+export function renderFilterByType(options: PaginationOptions): void {
+  const typeOptionHTML = FILTER_TYPE_OPTIONS.map(item => `<option value="${item.title}">${item.title}</option>`).join('');
+  if (selectType == null) {
+    return;
+  }
+  selectType.innerHTML = typeOptionHTML;
+  setDefaultSelected(selectType, FILTER_TYPE_OPTIONS.filter(item =>
+    item.value === LocalStorageService.getValue<TypeModel>(KEY_TYPE))[0]?.title ??
+    FILTER_TYPE_OPTIONS[0].title);
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderFilterByType.ts:25]:
```suggestion
  setDefaultSelected(selectType, FILTER_TYPE_OPTIONS.filter(item =>
    item.value === LocalStorageService.getValue<TypeModel>(KEY_TYPE))[0]?.title ??
    FILTER_TYPE_OPTIONS[0].title
  );
```

File: apps/vanilla/src/scripts/renderPagination.ts [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    const animeList = await renderAnimeList(options);
+    const optionUpdated = new PaginationOptions({
+      ...options,
+      totalPages: Math.ceil(animeList.count / options.limit - 1),
+    });
+    renderPagination(optionUpdated);
+    const itemsPageList = document.querySelectorAll('.pagination li:not(.btn)');
+    itemsPageList.forEach(item => {
+      item.addEventListener('click', () => {
+        const strPage = item.childNodes[0].childNodes[0].nodeValue;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderPagination.ts:106]: https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+import { LocalStorageService } from '../service/localStorage';
+
+import { renderListAnimeWithActivePage } from './renderPagination';
+
+const selectSort = document.querySelector<HTMLSelectElement>('.filter__item_select-sort');
+const selectOrdering = document.querySelector<HTMLSelectElement>('.filter__item_select-order');
+
+/** Init and render sorting and ordering list, hence render corresponding list anime.
+ * @param options Options of pagination.
+ */
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:17]: Please format **all** multiline jsdoc properly
```suggestion
/**
 * Init and render sorting and ordering list, hence render corresponding list anime.
 * @param options Options of pagination.
 */
```

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    return;
+  }
+  const sortOptionHTML = SORT_OPTIONS.map(item => (
+    `<option value="${item.title}">${item.title}</option>`
+  )).join('');
+
+  const orderOptionHTML = Object.keys(OrderOption).map(item => (
+    `<option value ="${item}">${item}</option>`
+  ))
+    .join('');
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:29]: Be consistent
```suggestion
  const sortOptionHTML = SORT_OPTIONS.map(item => (
    `<option value="${item.title}">${item.title}</option>`
  )).join('');

  const orderOptionHTML = Object.keys(OrderOption).map(item => (
    `<option value ="${item}">${item}</option>`
  )).join('');
```

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  selectSort.innerHTML = sortOptionHTML;
+  setDefaultSelected(selectSort, LocalStorageService.getValue<Sorting>(KEY_SORTING)?.title ?? SORT_OPTIONS[0].title);
+  selectSort.addEventListener('change', () => {
+    const { value } = selectSort;
+    LocalStorageService.setValue(KEY_SORTING, SORT_OPTIONS.filter(item => item.title === value)[0]);
+    const selectSortingValue = LocalStorageService.getValue<Sorting>(KEY_SORTING) ?? SORT_OPTIONS[0];
+    const optionsUpdated = new PaginationOptions({
+      ...options,
+      offset: DEFAULT_LIMIT,
+      activePage: 1,
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:39]: Avoid "magic numbers", use constant

File: libs/core/enum/anime.ts [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+export enum Status {
+
+  // Airing = 'AIRING',
+  // Finished = 'FINISHED',
+  // NotYetAir = 'NOT_YET_AIRED',
+  AIRING = 'Airing',
+  FINISHED = 'Finished',
+  NOT_YET_AIRED = 'Not yet aired',
+
+  // UNDEFINED = 'Undefined',
```
Comment (artem-sotnichenko-saritasa) [inline libs/core/enum/anime.ts:23]: Don't store commented code in a repository

File: apps/vanilla/src/constants/filterTypeOptions.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+interface FilterType {
+
+  /** Title of options. */
+  title: string;
+
+  /** Value of options. */
+  value: Type;
+}
+
+import { Type } from '@js-camp/core/enum';
```
Comment (timofei-maximov-saritasa) [inline apps/vanilla/src/constants/filterTypeOptions.ts:13]: Let's move it on top of file

File: apps/vanilla/src/constants/filterTypeOptions.ts [hunk 23]
```diff
@@ -0,0 +1,44 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+
+/** Options of type interface. */
+interface FilterType {
+
+  /** Title of options. */
+  title: string;
+
+  /** Value of options. */
+  value: Type;
```
Comment (timofei-maximov-saritasa) [inline apps/vanilla/src/constants/filterTypeOptions.ts:10]: Missing readonly

File: apps/vanilla/src/scripts/renderSortingAndOrdering.ts [hunk 24]
```diff
@@ -1,66 +1,71 @@
...
 const selectSort = document.querySelector<HTMLSelectElement>('.filter__item_select-sort');
 const selectOrdering = document.querySelector<HTMLSelectElement>('.filter__item_select-order');
 
-/** Init and render sorting and ordering list, hence render corresponding list anime.
+/**
+ * Init and render sorting and ordering list, hence render corresponding list anime.
  * @param options Options of pagination.
  */
 export function renderSortingAndOrdering(options: PaginationOptions): void {
+  if (selectSort === null || selectSort === undefined) {
```
Comment (timofei-maximov-saritasa) [inline apps/vanilla/src/scripts/renderSortingAndOrdering.ts:20]: `querySelector` returns `Element | null`. There is no need to check if the value is `undefined`

File: libs/core/utils/index.ts [hunk 25]
```diff
@@ -7,3 +7,20 @@ export function formatDate(dateToFormat: Date): string {
...
+  const optionList = selectElement.querySelectorAll('option');
+  optionList.forEach(option => {
+    if (defaultValue === null || defaultValue === undefined) {
+      return;
+    }
+    if (option.value === defaultValue) {
+      option.setAttribute('selected', 'true');
+    }
+  });
+}
```
Comment (timofei-maximov-saritasa) [inline libs/core/utils/index.ts:26]: I don't think this function belongs here. It won't be used anywhere but vanilla app.

File: apps/vanilla/src/scripts/renderFilterByType.ts [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+  selectType.addEventListener('change', () => {
+    LocalStorageService.setValue(KEY_TYPE, FILTER_TYPE_OPTIONS.filter(item => selectType.value === item.title)[0].value);
+    const valueType = LocalStorageService.getValue<Type>(KEY_TYPE) ?? Type.Default;
+    const optionsUpdated = new PaginationOptions({
+      ...options,
+      activePage: DEFAULT_ACTIVE_PAGE,
+      type: valueType,
+    });
+    renderListAnimeWithActivePage(optionsUpdated);
+  });
```
Comment (timofei-maximov-saritasa) [inline apps/vanilla/src/scripts/renderFilterByType.ts:35]: Let's improve code readability by separating parts of this code with empty lines

**Pull request closed without merging**